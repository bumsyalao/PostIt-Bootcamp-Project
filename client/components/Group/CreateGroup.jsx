import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { createGroupRequest } from '../../actions/createGroupRequest';
import { connect } from 'react-redux';
import Sidebar from '../Sidebar';
import NavigationBar from '../NavigationBar';

class CreateGroup extends Component {
    constructor(props) {
    super(props);
    this.state = {
      id:'',
      groupname: '',
      description: ''
    }
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }
  
  componentDidMount(){
    $('.collapsible').collapsible();
  }

  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }
  
  onSubmit(event) {
    event.preventDefault();
    console.log(this.state.groupname);
    this.props.createGroupRequest(this.state.groupname)
      .then(() => {
        Materialize.toast(created, 5000, 'green');
      }).catch((err) => {
        Materialize.toast(err, 5000, 'red');
      });
  }


  render() {
    const { createGroupRequest } = this.props;
    return (
      <div>
        <Sidebar />
        <div className="homepage">
          <NavigationBar />
          <div className="col s12 container">
            <div className="input-field col s6 offset-s3">
              <i className="material-icons prefix">group</i>
              <input value={this.state.groupname} onChange={this.onChange} name="groupname" type="text" 
              className="validate" required/>
              <label className="active" htmlFor="groupname">Groupname</label>
            </div>
            <div className="col s6 offset-s3">
              <label>Add User</label>
              <select className="browser-default">
                <option value="" disabled selected>Select a User</option>
                <option value="1">Option 1</option>
              </select>
            </div>
              <button onClick={this.onSubmit} disabled={this.state.invalid} className="btn waves-effect waves-light col s6 offset-s3 red lighten-2" type="submit" name="action">Enter
                <i className="material-icons right">send</i>
              </button>
          </div>
        </div>
      </div>
    )
  }
}

CreateGroup.propTypes = {
  createGroupRequest: PropTypes.func.isRequired,
};

export default connect (null, {createGroupRequest})(CreateGroup);
