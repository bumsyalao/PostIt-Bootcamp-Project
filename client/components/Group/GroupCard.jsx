import React, { Component } from 'react';
import { Link } from 'react-router-dom';

class GroupCard extends Component {
  /* eslint-disable */
  /**
   * Mounts materialize collapsible
   * @memberOf GroupCard
   */
  componentDidMount() {
    $('.collapsible').collapsible();
  }
  /* eslint-enable */
  /**
   * Renders GroupCard component
   * @returns GroupCard component
   * @memberOf GroupCard
   */
  render() {
    const { groupname, id, joinGroup, listUsers, users } = this.props;

    return (
      <div className="row">
        <div className="col s12 container">
          <div className="message-card group">
            <div className="card-action">
              <ul className="collapsible " data-collapsible="accordion">
                <li>
                  <div
                    className="collapsible-header group-header"
                    id="document-header"
                  >
                    <Link
                      to={`/homepage/view-group/${id}`}
                      className="card-title"
                    >
                      {groupname}
                    </Link>
                    <i className="material-icons">more_vert</i>
                  </div>
                  <div className="collapsible-body options" id="groups">
                    <ul
                      className="collapsible group-options"
                      data-collapsible="accordion"
                    >
                      <li>
                        <i className="material-icons tiny">games</i>
                        <div onClick={joinGroup} id={id} className="card-title">
                          Join group
                        </div>
                      </li>

                      <li className="user-list">
                        <div
                          className="collapsible-header group-header menu-item"
                          id="document-header"
                        >
                          <i className="material-icons tiny">group</i>
                          <div
                            onClick={() => listUsers(id)}
                            id={id}
                            className="card-title"
                          >
                            View users
                          </div>
                        </div>
                        <div className="collapsible-body options" id="groups">
                          <ul
                            className="collapsible group-options"
                            data-collapsible="accordion"
                          >
                            <li>
                              <div className="users-list">
                                {users &&
                                  users.map(({ username }) => (
                                    <div id={id} className="card-title">
                                      {username}
                                    </div>
                                  ))}
                              </div>
                            </li>
                          </ul>
                        </div>
                      </li>
                    </ul>
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
export default GroupCard;
