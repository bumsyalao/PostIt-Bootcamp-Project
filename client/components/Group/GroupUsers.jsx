import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { listAllUsers } from '../../actions/groups';


/**
 *
 *
 * @class GroupUsers
 * @extends {Component}
 */
export class GroupUsers extends Component {
  /**
   *
   * Makes action call to list all users in group
   * @memberOf GroupChat
   */
  componentWillMount() {
    const id = this.props.match.params.groupid;
    this.props.listAllUsers(id);
  }

  /**
   *
   * Renders GroupUsers component
   * @returns GroupUsers
   *
   * @memberOf GroupUsers
   */
  render() {
    const groupId = this.props.match.params.groupid;
    return (
    <div className="row">
            <Link
          to={`/homepage/view-group/${groupId}`}
          className="waves-effect waves-light red lighten-2 btn">
          {' '}
          Go Back to Group
        </Link>
        <div className="container form-margin">
          <div className="col s12 m7">
            <ul className="collection">
            {this.props.groupList[0] && this.props.groupList[0].users.map(user => (
               <li key ={user.userId} className="collection-item">
               {user.username}
              </li>
            ))}
            </ul>
          </div>
        </div>
      </div>
    );
  }
}
const mapStateToProps = state => ({
  groupList: state.group.groupList
});

export default connect(mapStateToProps, { listAllUsers })(GroupUsers);
