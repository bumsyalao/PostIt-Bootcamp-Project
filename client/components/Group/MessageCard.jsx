import React from 'react';

/**
 *
 *
 * @param {any} props 
 * @returns 
 */
const MessageCard = (props) => {
  const date = new Date(props.createdAt);
  return (
    <li className="collection-item">
      <div className="row">
        <div className="col s12 container">
          <div className="message-card">
            <div className="message-head">
              <div className="sender">
                <h6>{(props.username || props.User.username).toUpperCase()}</h6>
              </div>
              <div className="priority">
                Priority: &nbsp;
                <span>{props.messagePriority}</span>
              </div>
              <div className="time">
              {
                `${date.toDateString()} at ${
                  date.getHours()}:${date.getMinutes()}`
              }
              </div>
            </div>
            <div className="message-body"> {props.message}</div>
          </div>
        </div>
      </div>
    </li>
  );
};
export default MessageCard;
