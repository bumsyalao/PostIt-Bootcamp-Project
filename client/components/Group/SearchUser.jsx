import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import ReactPaginate from 'react-paginate';
import { getAllUsers } from '../../actions/users';
import { addMemberToGroup, getGroups } from '../../actions/groups';

/**
 * @class SearchPage
 * @extends {Component}
 */
export class SearchUser extends Component {
  /**
   * Creates an instance of SearchPage.
   * Bind the function handlePageClick to the class.
   * @memberOf SearchPage
   */
  constructor() {
    super();
    this.state = {
      users: [],
      offset: 0,
      pageCount: 0,
      limit: 5,
      searchParam: ''
    };
    this.joinGroup = this.joinGroup.bind(this);
    this.handlePageClick = this.handlePageClick.bind(this);
  }

  /**
   * Makes action call to get all users
   * @memberOf SearchPage
   */

  componentWillMount() {
    this.props.getGroups();
    this.getUsers();
  }

  /**
   * function to get users of the offset, limit and serach param
   * @param {int} offset
   * @param {int} limit
   * @memberOf SearchPage
   */

  getUsers(offset, limit) {
    this.props.getAllUsers(offset, limit, this.state.searchParam);
  }

  /**
   * Update the state if the props are changed
   * @param {object} nextProps
   * @memberOf SearchPage
   */
  componentWillReceiveProps(nextProps) {
    this.setState({
      users: nextProps.users.users,
      pageCount: nextProps.pagination.pageCount,
      count: nextProps.pagination.count
    });
  }
  /**
   *
   * Makes an action call to addMemberToGroup
   * @param {object} event The event of the HTML component
   *
   * @memberOf SearchUser
   */

  joinGroup(event) {
    event.preventDefault();
    const id = event.target.id;
    const groupId = this.props.match.params.groupId;
    this.props
      .addMemberToGroup(groupId, id)
      .then(() => {
        Materialize.toast('Member successfully added', 5000, 'green');
      })
      .catch(({ response }) => {
        Materialize.toast(
          `An error occured: ${response.data.message}`,
          5000,
          'red'
        );
      });
  }

  /**
   * Sets the event value and default value to the state
   * @function searchUsers
   * @param {object} event
   * @memberOf SearchPage
   */
  searchUsers = (event) => {
    this.setState({ searchParam: event.target.value, offset: 0, limit: 5 });
  };

  /**
   * Makes an action call to search for a user
   * @function onSearch
   * @memberOf SearchPage
   */
  onSearch = () => {
    const limit = 5,
      offset = 0;
    if (!this.state.searchParam) return;
    this.getUsers(offset, limit);
  };

  /**
   * Pagination for list of users
   * @param {object} data
   * @memberOf SearchPage
   */
  handlePageClick(data) {
    const selected = data.selected;
    const limit = 5;
    const offset = Math.ceil(selected * limit);
    this.setState({ offset });
    this.getUsers(offset, limit);
  }

  /**
   * Renders the SearchPage document
   * @returns SearchPage
   * @memberOf SearchPage
   */
  render() {
    const { users } = this.state;
    const groupid = this.props.match.params.groupId;
    return (
      <div className="row">
        <Link
          to={`/homepage/view-group/${groupid}`}
          className="waves-effect waves-light red lighten-2 btn"
        >
          {' '}
          Go Back to Group
        </Link>
        <div className="container form-margin">
          <div className="">
            <div className="row">
              <form className="col s12">
                <div className="row">
                  <div className="input-field col s12">
                    <i className="material-icons prefix">search</i>
                    <input
                      id="icon_prefix"
                      type="text"
                      onChange={this.searchUsers}
                      className="validate"
                    />
                    <label htmlFor="icon_prefix">Enter username</label>
                    <a
                      className="waves-effect waves-light red lighten-2 btn"
                      onClick={this.onSearch}
                    >
                      Search
                    </a>
                  </div>
                </div>
              </form>
            </div>
          </div>
        </div>

        <div className="container table">
          <table className="responsive-table">
            <thead>
              <tr>
                <th>USERNAME</th>
                <th>EMAIL</th>
              </tr>
            </thead>
            <tbody>
              {users &&
                users.map(user => (
                  <tr key={user.id}>
                    <td>
                      <i
                        className="material-icons icon-pointer"
                        onClick={this.joinGroup}
                        id={user.id}
                      >
                        add_box
                      </i>
                      {user.username}
                    </td>
                    <td>{user.email}</td>
                  </tr>
                ))}{' '}
            </tbody>
          </table>
          {this.state.count > 5 && (
            <ReactPaginate
              previousLabel={'previous'}
              nextLabel={'next'}
              pageCount={this.state.pageCount}
              marginPagesDisplayed={1}
              pageRangeDisplayed={3}
              onPageChange={this.handlePageClick}
              containerClassName={'pagination'}
              subContainerClassName={'pages pagination'}
              activeClassName={'active'}
            />
          )}
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  users: state.users,
  pagination: state.users.pagination,
  groupList: state.group.groupList
});

export default connect(mapStateToProps, {
  getAllUsers,
  addMemberToGroup,
  getGroups
})(SearchUser);
