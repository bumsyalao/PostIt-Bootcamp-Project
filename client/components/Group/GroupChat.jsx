import React from "react";
import { connect } from 'react-redux';
import MessageCard from './MessageCard';
import { newMessage, getMessages } from '../../actions/messages';

class GroupChat extends React.Component {
  constructor(props){
    super(props)
      this.state = {
        message: '',
        messagePriority: ''
      }
    this.onSubmit = this.onSubmit.bind(this);
    this.onChange = this.onChange.bind(this);
  }
  componentDidMount() {
    $('select').material_select();
    const { groupId } = this.props.match.params;
    this.props.getMessages(groupId);
  
  }

  onChange(event) {
    this.setState ({
      [event.target.name]: event.target.value});

  }
  optionChange(event) {
    this.setState ({ messagePriority: event.target.value});

  }
  onSubmit() {
    const { message, messagePriority } = this.state;
    const { groupId } = this.props.match.params;
    this.props.newMessage(groupId, this.props.username, { message, messagePriority })
      .then(() => {
        console.log(this.message);
      })
  }

  render() {
    console.log(this.props.messages);
    return (
    <div>
      <div className="row">
        <div className="col s12 m5 l5">
          {this.props.messages.map((message) => 
             <MessageCard key={message.id} {...message} />)
          }
           <div className="message-card2 ">
            <label for="textarea">Enter Message Here</label>
            <textarea value={this.message} name="message" id="textarea" onChange={this.onChange} />
            <label>Priority</label>
            <select
              name="messagePriority"
              className="browser-default input-field select" onChange={this.onChange}>
              <option value="">Choose Message Priority</option>
              <option value="normal">Normal</option>
              <option value="urgent">Urgent</option>
              <option value="critical">Critical</option>
            </select>
            <button onClick={this.onSubmit}> Send </button>
          </div>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  const { groupId } = ownProps.match.params;
  console.log(state.messages)
  return {
    username: state.access.user.username,
    messages: state.messages[groupId] || []
  }
}

export default connect(mapStateToProps, {newMessage, getMessages})(GroupChat);

