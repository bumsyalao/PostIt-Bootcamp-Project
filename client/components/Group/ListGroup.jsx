import React from 'react';
import { connect } from 'react-redux';
import { Switch, Route, withRouter } from 'react-router-dom';
import { groupsRequest } from '../../actions/createGroupRequest';
import { getMessages } from '../../actions/messages';
import { getGroups } from '../../actions/groups';
import GroupCard from './GroupCard';
import GroupChat from './GroupChat';


class ListGroup extends React.Component {
  constructor(props){
    super(props)
    this.onClick = this.onClick.bind(this);
    this.joinGroup = this.joinGroup.bind(this);
  }
  
  componentDidMount(){
    this.props.getGroups();
  }

  onClick(id){
        this.props.history.push(`/homepage/view-group/${id}`)
  }

  joinGroup (id) {
    // event.preventDefault();
    // const id = event.target.data-id;
    console.log("something", id);
    // this.props.getMessages(id).then((err, res) => {
    //   if(!err) {
    //     console.log("error when no error");
    //   }
    //   console.log("something when error");
    // });
  }

  render () {
    return (
      <div>
        <div className="row">
          <div className="col s12 m5 l5">
             {this.props.groupList.map((group) => 
             <GroupCard key={group.id} id={group.id} onClick={() => 
               this.onClick(group.id)} joinGroup={this.joinGroup} {...group} />)}
          </div>
          <div className="col s12 m6 l6">
            {/* <Switch>
              <Route path='/homepage/view-group' component={GroupChat}/> 
            </Switch>  */}
          </div>
        </div>
      </div>
    )}
}

const mapStateToProps = state => (
  {
    group: state.group.group,
    groupList: state.group.groupList,
    messages: state.group.groupMessages
  }
);

export default connect(mapStateToProps, {getGroups, groupsRequest, getMessages})(withRouter(ListGroup));