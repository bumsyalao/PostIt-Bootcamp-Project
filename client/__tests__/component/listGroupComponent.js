/* global expect jest test */
import React from 'react';
import { shallow, configure, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import adapter from 'enzyme-adapter-react-16';
import { ListGroup } from '../../components/Group/ListGroup';
import { GroupCard } from '../../components/Group/GroupCard';

configure({ adapter: new adapter() });

describe('ListGroup component', () => {
  const spy = jest.fn();
  beforeEach(() => {
    global.Materialize = { toast: jest.fn() };
  });
  afterEach(() => {
    spy.mockReset();
  });
  const props = {
    history: {
      push: spy
    },
    groupList: {
      map: spy
    },
    groups: [{
      id: 'vgh',
      groupname: 'Bunmi'
    }],
    getGroups: jest.fn(() => Promise.resolve()),
    addMemberToGroup: jest.fn(() => Promise.resolve()),
    listAllUsers: jest.fn(() => Promise.resolve())
  };
  const component = shallow(
    <ListGroup {...props} />
  );
  const childComponent = shallow(
    <GroupCard {...props} />
  );
  test('should match the ListGroupSnapshot', () => {
    const tree = toJson(component);
    expect(component.find('GroupCard').length).toEqual(0);
    expect(tree).toMatchSnapshot();
  });

  test('it should run function onClick', () => {
    const newerspy = jest.spyOn(component.instance(), 'joinGroup');
    component.find('#id').simulate('click');
    expect(newerspy).toHaveBeenCalled();
  });
});
