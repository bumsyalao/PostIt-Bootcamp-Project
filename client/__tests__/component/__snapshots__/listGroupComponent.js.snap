// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListGroup component should match the ListGroupSnapshot 1`] = `
<div>
  <div
    className="row form-margin"
  >
    <div
      className="col s12 m12 l12"
    />
    <div
      className="col s12 m12 l12"
    />
  </div>
</div>
`;
